[DynastyScans] Use JSON API

medmig opened the pull request:
Use JSON API instead of scraping HTML.

Commit:
[DynastyScans] Use JSON API
 --> Reduce server load and loading time

Commit:
Merge branch 'master' of https://github.com/manga-download/hakuneko.git into DynastyScans_Json

27136773 commented on src/web/mjs/connectors/DynastyScans.mjs:
```diff
@@ -26,14 +26,17 @@ export default class DynastyScans extends Connector {
-                            title: element.text.trim()
-                        };
-                    } );
+            return this.fetchJSON( this.url + `${page}.json` )
```
Tip: In the current HakuNeko versions you can also use async/await pattern, which makes the promises more readable.
Example: https://github.com/manga-download/hakuneko/wiki/Developer-Website-Tutorial#47-from-callbacks-over-promises-to-asyncawait

medmig replied:
Noted, thanks.
I won't have time to make the modifications tonight, I'll do it ASAP and push again !

27136773 commented on src/web/mjs/connectors/DynastyScans.mjs:
```diff
@@ -52,12 +55,13 @@ export default class DynastyScans extends Connector {
+            .then( json => {
+                // filter removes the "header" elements from the list
+                let chapterList = json['taggings'].filter( chapter => !chapter.header ).map( chapter => {
+                    let title = chapter.title.replace( manga.title, '' ).trim();
```
Does the json data still has the manga title within the chapter title and needs to be replaced?

medmig replied:
Maybe not, it's the same as in HTML pages. Actually I don't really remember why it was like that in the first place.

medmig replied:
Quickly searched but didn't find any example where this would apply.
However I found [an example](https://dynasty-scans.com/anthologies/an_absurd_relationship_collection) of why it might have been done this way. The same series is also under another format [here](https://dynasty-scans.com/series/an_absurd_relationship).
I'll simplify it since it can easily be restored later if needed anyway.

EDIT: Found [an example](https://dynasty-scans.com/anthologies/mermaid_line) where it was applicable. Still, not sure it is really important or bothering anyway...

27136773 commented on src/web/mjs/connectors/DynastyScans.mjs:
```diff
@@ -82,16 +86,9 @@ export default class DynastyScans extends Connector {
-                let pageList = JSON.parse( pages ).map( p => this.url + p.image );
+        this.fetchJSON( this.url + chapter.id + '.json' )
+            .then( json => {
+                let pageList = json['pages'].map( page => this.url + page.url );
```
May use `new URL(page.url, this.url).href`, this will cover the case that page.url already is an absolute link

medmig replied:
Good idea, thanks. I'll update this ASAP too.

27136773 reviewed (approved):
Not tested, but source looks good

Commit:
[DynastyScans] Minor corrections and improvements
 - Correct an issue where a "chapter" could be a header with value "null" (e.g. /anthologies/cider_and_crybaby)
